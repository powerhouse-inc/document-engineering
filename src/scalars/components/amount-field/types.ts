import type { AmountInputProps } from "../../../ui/components/data-entry/amount-input/amount-input.js";
import type { FieldErrorHandling } from "../types.js";

type AmountFieldProps = AmountInputProps & FieldErrorHandling;

export type { AmountFieldProps };
