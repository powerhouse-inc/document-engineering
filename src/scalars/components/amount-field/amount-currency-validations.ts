import type { Amount } from '../../../ui/components/data-entry/amount-input/types.js'
import type { AmountFieldProps } from './types.js'

export const validateAmountCurrency =
  ({ type, required }: AmountFieldProps) =>
  (value: unknown) => {
    if (!value) {
      return required ? 'Please select a valid currency' : true
    }
    if (['AmountFiat', 'AmountCrypto', 'AmountCurrency'].includes(type)) {
      const actualUnit = (value as Amount).unit
      return required && !actualUnit ? 'Please select a valid currency' : true
    }

    if (type === 'Amount' && typeof value === 'object' && 'unit' in value) {
      const actualUnit = (value as Amount).unit
      return required && !actualUnit ? 'Please select a valid currency' : true
    }

    return true
  }
