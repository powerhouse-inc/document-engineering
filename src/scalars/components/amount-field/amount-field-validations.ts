import type { ValidatorResult } from '../../../scalars/components/types.js'
import type {
  Amount,
  AmountCrypto,
  AmountCurrency,
  AmountFiat,
  AmountInputPropsGeneric,
  AmountValue,
} from '../../../ui/components/data-entry/amount-input/types.js'
import { isValidBigInt } from '../../../ui/components/data-entry/amount-input/utils.js'
import { isValidNumber } from '../number-field/number-field-validations.js'
import type { AmountFieldProps } from './types.js'

const isAmountCurrencyFiat = (type: AmountInputPropsGeneric['type']): type is 'AmountFiat' => type === 'AmountFiat'

const isAmountCurrencyCrypto = (type: AmountInputPropsGeneric['type']): type is 'AmountCrypto' =>
  type === 'AmountCrypto'

const isAmountCurrencyUniversal = (type: AmountInputPropsGeneric['type']): type is 'AmountCurrency' =>
  type === 'AmountCurrency'

const isAmount = (type: AmountInputPropsGeneric['type']): type is 'Amount' => type === 'Amount'

const getAmount = (value: AmountValue, type: AmountInputPropsGeneric['type']): number | string | undefined => {
  if (isAmountCurrencyFiat(type) || isAmountCurrencyCrypto(type) || isAmountCurrencyUniversal(type) || isAmount(type)) {
    if (!value) return undefined
    const amount = (value as AmountFiat | AmountCrypto | AmountCurrency | Amount).amount ?? undefined
    return amount
  }
  return value as number
}

export const validateAmount =
  ({ type, minValue, maxValue, allowNegative, required }: AmountFieldProps) =>
  (value: unknown): ValidatorResult => {
    const amount = getAmount(value as AmountValue, type)
    if (amount === '' || amount === undefined) {
      return required ? 'Please enter a valid number' : true
    }

    // This case are types that are strings but hndle as bigint
    if (type === 'AmountCurrency' || type === 'AmountCrypto') {
      if (!isValidBigInt(amount.toString())) {
        return 'Value is not an bigint'
      }

      if (!allowNegative && BigInt(amount) < 0) {
        return 'Value must be positive'
      }
      if (maxValue) {
        if (BigInt(amount) > BigInt(maxValue)) {
          return `This field must be less than ${maxValue}`
        }
      }
      if (minValue) {
        if (BigInt(amount) < BigInt(minValue)) {
          return `This field must be more than ${minValue}`
        }
      }
      return true
    }

    if (Math.abs(Number(amount.toString())) > Number.MAX_SAFE_INTEGER) {
      return 'Value is too large for number'
    }
    if (!isValidNumber(amount)) {
      return 'Value is not a valid number'
    }
    if (!allowNegative && Number(amount) < 0) {
      return 'Value must be positive'
    }
    if (maxValue) {
      if (Number(amount) > maxValue) {
        return `This field must be less than ${maxValue}`
      }
    }
    if (minValue) {
      if (Number(amount) < minValue) {
        return `This field must be more than ${minValue}`
      }
    }
    return true
  }
