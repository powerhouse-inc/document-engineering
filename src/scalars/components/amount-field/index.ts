export { AmountField } from './amount-field.js'
export type { AmountFieldProps } from './types.js'
