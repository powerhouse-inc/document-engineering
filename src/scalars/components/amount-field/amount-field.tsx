'use client'

import { AmountInput } from '../../../ui/components/data-entry/amount-input/amount-input.js'
import { withFieldValidation } from '../fragments/with-field-validation/with-field-validation.js'
import type { AmountFieldProps } from './types.js'
import { validateAmountCurrency } from './amount-currency-validations.js'
import { validateAmount } from './amount-field-validations.js'

const AmountField = withFieldValidation<AmountFieldProps>(AmountInput, {
  validations: {
    _numericAmount: validateAmount,
    _validOptionCurrency: validateAmountCurrency,
  },
})

AmountField.displayName = 'AmountField'

export { AmountField }
