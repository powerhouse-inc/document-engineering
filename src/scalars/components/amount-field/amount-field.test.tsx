import {
  commonCryptoCurrencies,
  commonFiatCurrencies,
} from '../../../ui/components/data-entry/currency-code-picker/defaults.js'
import { screen, waitFor } from '@testing-library/react'
import { renderWithForm } from '../../lib/testing.js'
import { AmountField } from './amount-field.js'
import userEvent from '@testing-library/user-event'

describe('AmountField Component', () => {
  it('should match snapshot', () => {
    const { container } = renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          value: 345,
        }}
        units={commonCryptoCurrencies}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render label when provided', () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          value: 345,
        }}
        units={commonCryptoCurrencies}
      />
    )
    expect(screen.getByLabelText('Amount Label')).toBeInTheDocument()
  })

  it('should render error messages when provided', async () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          value: 345,
        }}
        errors={['Error 1', 'Error 2']}
        validators={() => 'Error 3'}
        units={commonCryptoCurrencies}
      />
    )
    await waitFor(() => {
      expect(screen.getByText('Error 1')).toBeInTheDocument()
      expect(screen.getByText('Error 2')).toBeInTheDocument()
      expect(screen.getByText('Error 3')).toBeInTheDocument()
    })
  })

  it('should render the percentage sign if the type is percent', () => {
    renderWithForm(<AmountField label="Amount Label" name="amount" type="AmountPercentage" value={345} step={0} />)
    expect(screen.getByText('%')).toBeInTheDocument()
  })

  it('should disable the input when disabled prop is true', () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          value: 345,
        }}
        units={commonCryptoCurrencies}
        disabled
      />
    )
    const input = screen.getByLabelText('Amount Label')

    expect(input).toBeDisabled()
  })

  it('should set the input as required when required prop is true', () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          value: 345,
        }}
        required
        units={commonCryptoCurrencies}
      />
    )
    expect(screen.getByRole('spinbutton')).toHaveAttribute('required')
  })

  it('should render the description when provided', () => {
    renderWithForm(
      <AmountField
        label="Amount Label"
        name="amount"
        type="Amount"
        value={{
          value: 345,
        }}
        units={commonCryptoCurrencies}
        disabled
        description="This is a description"
      />
    )
    expect(screen.getByText('This is a description')).toBeInTheDocument()
  })

  it('should show validation error when currency is not selected', async () => {
    renderWithForm(
      <AmountField
        required
        label="Amount Label"
        name="amount"
        type="AmountFiat"
        showErrorOnBlur
        value={{
          value: 345,
          unit: '',
        }}
        units={commonFiatCurrencies}
        description="This is a description"
      />
    )

    const numberInput = screen.getByRole('spinbutton')
    await userEvent.click(numberInput)
    await userEvent.tab()
    expect(screen.getByText('Please select a valid currency')).toBeInTheDocument()
  })
})
