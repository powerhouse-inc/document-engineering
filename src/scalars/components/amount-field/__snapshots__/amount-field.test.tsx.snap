// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AmountField Component > should match snapshot 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:"
      >
        Amount Label
      </label>
      <div
        class="relative flex items-center rounded-md border border-gray-300 h-[36px] focus:rounded-md focus:border-gray-900 transition-colors focus-within:ring-1 focus-within:rounded focus-within:ring-offset-0"
      >
        <input
          data-cast="AmountNumber"
          name="amount"
          type="hidden"
        />
        <div
          class="relative flex items-center"
        >
          <div
            class="flex flex-col gap-2"
          >
            <div
              class="relative flex items-center group"
            >
              <input
                aria-invalid="false"
                class="flex w-full text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-gray-900 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 pr-8 border-r-[0.5px] rounded-md outline-none focus-visible:ring-0 focus-visible:ring-offset-0 border-none h-[32px]"
                data-cast="Number"
                id=":r1:"
                inputmode="numeric"
                name=""
                role="spinbutton"
                type="text"
                value="345"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
